impl_core/src/main/java/io/opencensus/implcore/stats/MeasureMapImpl.java

```diff
@@ -37,13 +41,21 @@ private MeasureMapImpl(StatsManager statsManager) {
 
   @Override
   public MeasureMapImpl put(MeasureDouble measure, double value) {
-    builder.put(measure, value);
+    if (value < 0) {
+      logger.log(Level.WARNING, "Dropping (" + value + "), value to record must be non-negative.");
+    } else {
+      builder.put(measure, value);
+    }
     return this;
   }
 
   @Override
   public MeasureMapImpl put(MeasureLong measure, long value) {
-    builder.put(measure, value);
+    if (value < 0) {
```

Comment:
Can we check and record warnings in the noop implementation as well?

I would not do this in Put, instead I would do it in the record. As I mentioned is better to drop all recorded values because users may use correlation between different values recorded at the same time.